### DISCUSSION TITLE
How to call uuid_generate_v4() as a default value for id in pq?

### DISCUSSION THREAD

**Discussion by viktor-ku** [sequelize/umzug#405] [Q&A]
I would greatly appreciate the answer!
Right now it seems like I cannot do that? There is `defaultValue`, but whatever string it accepts it just passes it like escaped string?

Same is for the dates, but I guess that is the same kind of question.

**So, how to call a function to set a default value?**

UPD. found `escape` and it looks like something I would use, but for pq it's set to `undefined`

**Comment by mmkal** [comment]
Hi - I need a bit more context here. Which default value? Which id? What is `pq`?? What's `escape`?

**Reply by viktor-ku** [reply]
When writing migration, you have `context`, which has various methods like `createTable` and others. For example, one of these methods is `escape`. Anyway, you have to specify types and in there you've got `type`, `primaryKey`, `allowNull` and others. You also have `defaultValue`, which I assume setups the default value, however, it doesn't work when writing a function call there like this:

```ts
export async function up({context: cx}: {context: QueryInterface}) {
  await cx.createTable(table, {
    id: {
      type: UUID,
      primaryKey: true,
      defaultValue: 'uuid_generate_v4()'
    },
  })
}
```

So my question is: how to generate a default UUID on the db side of things when inserting a new record? And `pq` is postgres. 

Sorry, for the lack of context, my bad.

Thanks

**Comment by viktor-ku** [comment]
For anyone wondering I just passed _default_ expression with the type info, which arguably feels very wrong, but hey

```ts
export async function up({context: cx}: {context: QueryInterface}) {
  await cx.createTable(table, {
    id: {
      type: 'uuid DEFAULT uuid_generate_v4()',
      primaryKey: true,
    },
  })
}

```

**Reply by mmkal** [reply]
Did you try using Sequelize.UUID? Something like the uuid example in https://sequelize.readthedocs.io/en/2.0/api/datatypes/

Note: I don't think this is to do with umzug, but sequelize.

**Reply by viktor-ku** [reply]
Link is 404.

I am not using js as a backend, so no, I am not using sequelize. I choose to run migrations in js because js is dynamic language and it allows me to manipulate data between migrations if need be.

**Reply by mmkal** [reply]
Have updated the link. It does look like the `cx` object is a sequelize queryInterface even if you don't use it in your app to talk to the db. Could you share your umzug constructor?

**Reply by viktor-ku** [reply]
Sure, it's pretty basic

```ts
  const db = new Sequelize(APP_DATABASE, {})
  // ...
  const migrations = new Umzug({
    migrations: {
      glob: `${migrationsDir}/*.ts`
    },
    context: db.getQueryInterface(),
    storage: new SequelizeStorage({sequelize: db}),
    logger: console
  })
```

Hmm, so you are saying that by specifying table inside of the migration, I actually am using this construction under the hood? Is that right?

```js
sequelize.define('model', {
  uuid: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV1,
    primaryKey: true
  }
})
```

**Reply by mmkal** [reply]
It has a similar interface I believe yes, though the queryInterface is lower-level. Docs here: https://sequelize.org/master/manual/query-interface.html

Tbh I'm not an expert on sequelize, I don't use it much outside of maintenance of this repo. Umzug aims to be database/framework agnostic, and just-so-happens to support sequelize out of the box.

Hopefully DataTypes can solve your problem here, but if you don't use sequelize outside of migrations and only are using it for dynamic migrations, it might be worth considering [@slonik/migrator](https://npmjs.com/package/@slonik/migrator), which is postgres-only migration tool which promotes writing sql rather than having to learn about all the abstractions of sequelize and how they map to database queries.

**Reply by viktor-ku** [reply]
Thank you for your help! I will be researching this topic. I see that the migrator prefers sql files, but I really really need to be able to manipulate data in between my migrations. Maybe I am doing something wrong, or I am weird, but I do `UPDATE` quite a lot in migrations too and it often requires many queries to db. Maybe you know what I am doing wrong?

**Reply by mmkal** [reply]
Script files are allowed too, that's fine. Personally I usually prefer sql because it makes it easier to know for sure what has run, and reduces lock-in to any given tool. But migrations can be complex so it's not necessarily a bad thing if js/ts is needed. Slonik is a typescript postgres client, it doesn't have multi-dialect support like sequelize. But if you're just using postgres that might not be a problem.

**Reply by viktor-ku** [reply]
Hey, thank you, I think I will give migrator a try, it sounds reasonable. Indeed, 90% of the time I just need to run sql.
